server/channels/app/channel_category.go

```diff
@@ -158,54 +191,74 @@ func (a *App) UpdateSidebarCategories(c request.CTX, userID, teamID string, cate
+func (a *App) muteChannelsForUpdatedCategories(c request.CTX, userID string, teamID string, modifiedCategories []*model.SidebarCategoryWithChannels, originalCategories []*model.SidebarCategoryWithChannels) {
+	anyCategoryMuted := func() bool {
+		for _, category := range append(modifiedCategories, originalCategories...) {
+			if category.Muted {
+				return true
+			}
+		}
+
+		return false
+	}
+
+	if !anyCategoryMuted() {
+		// None of the categories are/were muted, so no channels will be muted or unmuted
+		return
+	}
+
 	var channelsToMute []string
 	var channelsToUnmute []string
 
 	// Mute or unmute all channels in categories that were muted or unmuted
-	for i, updatedCategory := range updatedCategories {
-		if i > len(originalCategories)-1 {
-			// The two slices should be the same length, but double check that to be safe
-			continue
-		}
-
-		originalCategory := originalCategories[i]
-
-		if updatedCategory.Muted && !originalCategory.Muted {
-			channelsToMute = append(channelsToMute, updatedCategory.Channels...)
-		} else if !updatedCategory.Muted && originalCategory.Muted {
-			channelsToUnmute = append(channelsToUnmute, updatedCategory.Channels...)
+	originalCategoriesById := makeCategoryMap(originalCategories)
+	for _, modifiedCategory := range modifiedCategories {
+		originalCategory := originalCategoriesById[modifiedCategory.Id]
+
+		if modifiedCategory.Muted && !originalCategory.Muted {
+			channelsToMute = append(channelsToMute, modifiedCategory.Channels...)
+		} else if !modifiedCategory.Muted && originalCategory.Muted {
+			channelsToUnmute = append(channelsToUnmute, modifiedCategory.Channels...)
 		}
 	}
 
 	// Mute any channels moved from an unmuted category into a muted one and vice versa
-	channelsDiff := diffChannelsBetweenCategories(updatedCategories, originalCategories)
-	if len(channelsDiff) != 0 {
-		makeCategoryMap := func(categories []*model.SidebarCategoryWithChannels) map[string]*model.SidebarCategoryWithChannels {
-			result := make(map[string]*model.SidebarCategoryWithChannels)
-			for _, category := range categories {
-				result[category.Id] = category
+	anyChannelMoved := func() bool {
+		// If no channels were moved, we don't need to get the categories again to find where they were moved to/from
+		for _, modifiedCategory := range modifiedCategories {
+			if !slices.Equal(modifiedCategory.Channels, originalCategoriesById[modifiedCategory.Id].Channels) {
+				return true
 			}
-
-			return result
 		}
 
-		updatedCategoriesById := makeCategoryMap(updatedCategories)
-		originalCategoriesById := makeCategoryMap(originalCategories)
+		return false
+	}
 
-		for channelID, diff := range channelsDiff {
-			fromCategory := originalCategoriesById[diff.fromCategoryId]
-			toCategory := updatedCategoriesById[diff.toCategoryId]
+	if anyChannelMoved() {
+		updatedCategories, err := a.Srv().Store().Channel().GetSidebarCategoriesForTeamForUser(userID, teamID, true)
```

Comment:
I'm not a huge fan of adding the `fromMaster` parameter to this, but I think we need it in this case to avoid read-after-write bugs. Part of the reason that the previous `UpdateSidebarCategories` code returned `originalCategories` was to avoid those bugs, but this seems cleaner and more efficient even if it requires more store calls when a user has muted categories

Reply:
Thanks! I'll look into that.

FYI, I did notice that there's a few other places where we have parameters or separate methods to act on master instead of a replica

Reply:
Correct. All those should migrate to this method. Unfortunately we are behind in that aspect.